src/table/body-cell/inline-editor.tsx

```diff
@@ -0,0 +1,132 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useEffect, useState } from 'react';
+import Button from '../../button/internal';
+import FormField from '../../form-field/internal';
+import SpaceBetween from '../../space-between/internal';
+import { useClickAway } from './click-away';
+import { TableProps } from '../interfaces';
+import styles from './styles.css.js';
+import { Optional } from '../../internal/types';
+
+// A function that does nothing
+const noop = () => undefined;
+
+interface InlineEditorProps<ItemType> {
+  ariaLabels: TableProps['ariaLabels'];
+  column: TableProps.ColumnDefinition<ItemType>;
+  item: ItemType;
+  onEditEnd: () => void;
+  submitEdit: TableProps.SubmitEditFunction<ItemType>;
+  __onRender?: () => void;
+}
+
+export function InlineEditor<ItemType>({
+  ariaLabels,
+  item,
+  column,
+  onEditEnd,
+  submitEdit,
+  __onRender,
+}: InlineEditorProps<ItemType>) {
+  const [currentEditLoading, setCurrentEditLoading] = useState(false);
+  const [currentEditValue, setCurrentEditValue] = useState<Optional<any>>();
+
+  const cellContext = {
+    isEditing: true,
+    currentValue: currentEditValue,
+    setValue: setCurrentEditValue,
+  };
+
+  function finishEdit(cancel = false) {
+    if (!cancel) {
+      setCurrentEditValue(undefined);
+    }
+    onEditEnd();
+  }
+
+  async function onSubmitClick(evt: React.FormEvent) {
+    evt.preventDefault();
+    if (currentEditValue === undefined) {
+      finishEdit();
+      return;
+    }
+
+    setCurrentEditLoading(true);
+    try {
+      await submitEdit(item, column, currentEditValue);
+      setCurrentEditLoading(false);
+      finishEdit();
+    } catch (e) {
+      setCurrentEditLoading(false);
+    }
+  }
+
+  function onCancel() {
+    if (currentEditLoading) {
+      return;
```

Comment:
Good idea to have a test like "click away while loading does not cancel editable state" (or "pressing escape ..." or both)